src/backend/cuda/kernel/sift.hpp

```diff
@@ -0,0 +1,1306 @@
+#include <thrust/generate.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+#include <thrust/gather.h>
+#include <thrust/random.h>
+
+#include <cfloat>
+
+namespace cuda
+{
+
+namespace kernel
+{
+
+static const dim_t SIFT_THREADS   = 256;
+static const dim_t SIFT_THREADS_X = 32;
+static const dim_t SIFT_THREADS_Y = 8;
+
+static const float PI_VAL = 3.14159265358979323846f;
+
+// default width of descriptor histogram array
+static const int DescrWidth = 4;
+
+// default number of bins per histogram in descriptor array
+static const int DescrHistBins = 8;
+
+// assumed gaussian blur for input image
+static const float InitSigma = 0.5f;
+
+// width of border in which to ignore keypoints
+static const int ImgBorder = 5;
+
+// maximum steps of keypoint interpolation before failure
+static const int MaxInterpSteps = 5;
+
+// default number of bins in histogram for orientation assignment
+static const int OriHistBins = 36;
+
+// determines gaussian sigma for orientation assignment
+static const float OriSigFctr = 1.5f;
+
+// determines the radius of the region used in orientation assignment */
+static const float OriRadius = 3.0f * OriSigFctr;
+
+// number of passes of orientation histogram smoothing
+static const int SmoothOriPasses = 2;
+
+// orientation magnitude relative to max that results in new feature
+static const float OriPeakRatio = 0.8f;
+
+// determines the size of a single descriptor orientation histogram
+static const float DescrSclFctr = 3.f;
+
+// threshold on magnitude of elements of descriptor vector
+static const float DescrMagThr = 0.2f;
+
+// factor used to convert floating-point descriptor to unsigned char
+static const float IntDescrFctr = 512.f;
+
+template<typename T>
+void gaussian1D(T* out, const int dim, double sigma=0.0)
+{
+    if(!(sigma>0)) sigma = 0.25*dim;
+
+    T sum = (T)0;
+    for(int i=0;i<dim;i++)
+    {
+        int x = i-(dim-1)/2;
+        T el = 1. / sqrt(2 * PI_VAL * sigma*sigma) * exp(-((x*x)/(2*(sigma*sigma))));
+        out[i] = el;
+        sum   += el;
+    }
+
+    for(int k=0;k<dim;k++)
+        out[k] /= sum;
+}
+
+template<typename T>
+Param<T> gauss_filter(float sigma)
+{
+    // Using 6-sigma rule
+    unsigned gauss_len = (unsigned)round(sigma * 6 + 1) | 1;
+
+    T* h_gauss = new T[gauss_len];
+    gaussian1D(h_gauss, gauss_len, sigma);
+    Param<T> gauss_filter;
+    gauss_filter.dims[0] = gauss_len;
+    gauss_filter.strides[0] = 1;
+
+    for (int k = 1; k < 4; k++) {
+        gauss_filter.dims[k] = 1;
+        gauss_filter.strides[k] = gauss_filter.dims[k-1] * gauss_filter.strides[k-1];
+    }
+
+    dim_t gauss_elem = gauss_filter.strides[3] * gauss_filter.dims[3];
+    gauss_filter.ptr = memAlloc<T>(gauss_elem);
+    CUDA_CHECK(cudaMemcpy(gauss_filter.ptr, h_gauss, gauss_elem * sizeof(T), cudaMemcpyHostToDevice));
+
+    delete[] h_gauss;
+
+    return gauss_filter;
+}
+
+__inline__ __device__ void gaussianElimination(float* A, float* b, float* x, const int n)
+{
```

Comment:
Can you make `n` a template and call unroll on the loops below ? It looks like the value of `n` is known before hand for now.